### DISCUSSION TITLE
Codebehind for Oqtane-Component and Oqtane-Module

### DISCUSSION THREAD

**Discussion by CebiConRepo** [oqtane/oqtane.framework#1857] [Q&A]
## Hi there,
I'm trying to create a theme with codebehind files (for Oqtane). Unfortunately without success. It's not clear to me whether with or without partial. Does anyone have a little code sample for me? Best for one component and one module. It would be great if someone could help me.

**Comment by hishamco** [comment]
You can find some in https://github.com/oqtane/oqtane.framework/discussions/1852

**Reply by CebiConRepo** [reply]
Thanks for the information. Unfortunately there are no components or modules with code behind in the samples (blog, survey)! Or did I miss somethin
[Themes.zip](https://github.com/oqtane/oqtane.framework/files/7669151/Themes.zip)
g?

**Reply by hishamco** [reply]
Code-behind is a methodology, you can use code-behind or code in the component itself

**Reply by CebiConRepo** [reply]
if I remove "partial" it doesn't work either ..
![TestTheme-Razor-cs](https://user-images.githubusercontent.com/36736217/145050257-c620d4e8-6be5-4faa-a9bd-2ce553baceba.jpg)
![TestTheme-Razor](https://user-images.githubusercontent.com/36736217/145050275-8398f8f4-41b6-4127-a4cb-7315dc2a7f48.jpg)

.

**Reply by leigh-pointer** [reply]
I will take a look tomorrow and if possible creat a template package for both module and themes

**Reply by CebiConRepo** [reply]
That would be great, as an old webform developer I grew fond of codebehind. Just looks tidier. :-)

**Reply by leigh-pointer** [reply]
I have created a repo with and example of Theme using codebehind.  It hold no functionality other than what is default but it will set you up.

I will go on a create an repo with the Module template and from there I will create Templates.

Let me know if you require and help further.

https://github.com/leigh-pointer/MyCompany.MyThemeCodeBehind 

Enjoy

**Reply by leigh-pointer** [reply]
I have created a repo with and example of Module using codebehind. It holds no functionality other than what is default but it will set you up.

https://github.com/leigh-pointer/MyCompany.MyModuleCodeBehind

**Reply by CebiConRepo** [reply]
Great,
took a first look at your examples - just what I needed! Tonight I will build the templates into my project.

thanks

**Reply by leigh-pointer** [reply]
I will be uploading here a the module and the theme template files soon.  After a bit I will create a PR.

**Reply by leigh-pointer** [reply]
@CebiConRepo 
> Great, took a first look at your examples - just what I needed! Tonight I will build the templates into my project.
> 
> thanks

You are welcome

**Comment by sbwalker** [comment]
Just a note that this item was discussed last year in this thread https://github.com/oqtane/oqtane.framework/discussions/754 - at which time it was concluded that the framework itself would continue to use the single file approach. This decision was based primarily on the Microsoft Alignment and Consistency items in the Oqtane philosophy ( https://www.oqtane.org/Resources/Blog/PostId/538/oqtane-philosophy ). 

Certainly it is possible for developers to use the code behind approach for their own modules and themes, if that is their preference. It would also be possible for someone to create/maintain a code behind version of the templates which could be downloaded independently from the framework and placed in the appropriate folder locations (ie. Oqtane.Server\wwwroot\Modules\Templates) so that it can be used in the Module Creator.

**Reply by leigh-pointer** [reply]
> It would also be possible for someone to create/maintain a code behind version of the templates which could be downloaded independently from the framework and placed in the appropriate folder locations (ie. Oqtane.Server\wwwroot\Modules\Templates) so that it can be used in the Module Creator.

That sounds like a good approach, and so keeping with the Oqtane philosophy.  I will close the PR and move the templates to a repo on github.

**Reply by leigh-pointer** [reply]
I have created templates for the Standard Module and Theme that can be copied to your Oqtane Framework installation.
https://github.com/leigh-pointer/StuidoElf.Oqtane.Templates

**Reply by sbwalker** [reply]
FYI... your link is broken... however I should mention that if you want to make your templates even simpler to install for developers, you can create a simple module:

```
StudioElf.CodeBehindTemplate.Client
>> index.razor - this component does not have any actual functionality - it is just a placeholder to allow for install/uninstall - it just needs the right namespace and to inherit from ModuleBase
>> ModuleInfo.cs - IModule interface
StudioElf.CodeBehindTemplate.Package
>> the standard packaging assets
StudioElf.CodeBehindTemplate.Server
/wwwroot/Modules/Templates/StudioElf.CodeBehindTemplate
>> put your template assets in this folder
```
Package it up as an installable module. When a developer installs it, the installer will put the template assets in the correct folder so they can be used in the Module Creator. The same can be done for Themes.

**Reply by leigh-pointer** [reply]
@sbwalker thanks for that, actually the repo was created private but now is public.

I will create a packages for the Module and Theme templates, seems to be good way to go.

**Reply by leigh-pointer** [reply]
@sbwalker 
> Package it up as an installable module. When a developer installs it, the installer will put the template assets in the correct folder so they can be used in the Module Creator.

This works well in one direction.  It installs ok but when it come to uninstall it falls over in the InstallationManager.cs line 214                                 Directory.Delete(Path.GetDirectoryName(filepath));
 It complains about a file existing but it definitely doesn't as the directory is completely empty.  What is strange it doesn't report the first directory delete but the following.  The first directory delete does not delete but seems to skip the exception.

Appending ,true to the Directory.Delete fixes it ` Directory.Delete(Path.GetDirectoryName(filepath), true);`

**Reply by sbwalker** [reply]
@leigh-pointer can you post the related *.log file contents from the Oqtane.Server\Packages folder - it should provide a hint as to why the deletion logic is failing

**Reply by leigh-pointer** [reply]
@sbwalker 
[StudioElf.CodeBehindTemplates.1.0.0.log](https://github.com/oqtane/oqtane.framework/files/7693245/StudioElf.CodeBehindTemplates.1.0.0.log)

**Reply by sbwalker** [reply]
The behavior you are seeing is a bit strange... the uninstall logic reads the log file to discover the list of assets and processes them in reverse order so that it removes items lower in the directory tree first and works it way up the tree. The code:

```
                            if (!Directory.EnumerateFiles(Path.GetDirectoryName(filepath)).Any())
                            {
                                Directory.Delete(Path.GetDirectoryName(filepath));
                            }
```
is supposed to check if the folder contains any files - and only if it is empty,  delete it. However, you are saying the Directory.Delete is failing - which would seem to indicate there is a system or hidden file which the EnumerateFiles() method cannot see. Either way, I think the fix you suggested above is correct - please submit a PR when you get a chance. And thank you very much for testing both the install and uninstall aspects of this.

**Reply by leigh-pointer** [reply]
@sbwalker 
> please submit a PR when you get a chance. And thank you very much for testing both the install and uninstall aspects of this.
PR submitted.
